node/rpc/src/state_api.rs

```diff
@@ -209,72 +211,114 @@ pub(crate) async fn state_fetch_root<DB: Blockstore + Clone + Sync + Send + 'sta
                         log::debug!(
-                            "Still downloading. Fetched: {counter}, Failures: {failures}, Concurrent: {}",
-                            MAX_CONCURRENT_REQUESTS - sem.available_permits()
-                        );
+                                "Graph walk: CIDs: {counter}, Fetched: {fetched}, Failures: {failures}, dfs: {}, Concurrent: {}",
+                                dfs_guard.len(), task_set.len()
+                            );
+                    }
+                    if let Some(next_ipld) = db.get_cbor(&new_cid)? {
+                        dfs_guard.push(next_ipld);
+                    } else {
+                        to_be_fetched.push(new_cid);
                     }
-                    task_set.spawn({
-                        let network_send = data.network_send.clone();
-                        let db = data.chain_store.db.clone();
-                        let sem = sem.clone();
-                        async move {
-                            if !db.has(&new_cid)? {
-                                // If a CID isn't in our database, request it via bitswap (limited
-                                // by MAX_CONCURRENT_REQUESTS)
-                                let permit = sem.acquire_owned().await?;
-                                let (tx, rx) = flume::bounded(1);
-                                network_send
-                                    .send_async(NetworkMessage::BitswapRequest {
-                                        epoch: 0,
-                                        cid: new_cid,
-                                        response_channel: tx,
-                                    })
-                                    .await?;
-                                // Bitswap requests do not fail. They are just ignored if no-one has
-                                // the requested data. Here we arbitrary decide to only wait for
-                                // REQUEST_TIMEOUT before deciding that the data is unavailable.
-                                let _ignore = timeout(REQUEST_TIMEOUT, rx.recv_async()).await;
-                                drop(permit);
-                            }
-
-                            db.get_cbor::<Ipld>(&new_cid)?
-                                .ok_or_else(|| anyhow::anyhow!("Request failed: {new_cid}"))
-                        }
-                    });
                 }
             }
-            Err(msg) => {
-                failures += 1;
-                log::debug!("Request failed: {msg}");
+
+            while let Some(cid) = to_be_fetched.pop() {
+                if task_set.len() == MAX_CONCURRENT_REQUESTS {
+                    if let Some(ret) = task_set.join_next().await {
+                        handle_worker(&mut fetched, &mut failures, ret?)
+                    }
+                }
+                task_set.spawn_blocking({
```

Comment:
What's the reason for moving to `spawn_blocking` here?

Reply:
Pushing a value to the `dfs` mutex may block.

Reply:
But in that case the bitswap request is blocking as well now, no? Just wondering if it would not be better to have an async task calling `network_send.send_async` then pushing to the `dfs` mutex.

Reply:
Both ways are possible. Probably won't be any measurable difference.